### PR TITLE
Block maintenance operations during deletes

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorMetadata.java [hunk 1]
```diff
@@ -130,18 +131,26 @@
     private final MetadataDao dao;
     private final ShardManager shardManager;
     private final String connectorId;
+    private final Consumer<Long> deleteCallback;
```
Comment (dain) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorMetadata.java]: maybe name this `beginDeleteCallback` or `beginDeleteListener` or `beginDeleteMonitor`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 2]
```diff
@@ -76,6 +90,13 @@ public RaptorConnector(
         this.sessionProperties = requireNonNull(sessionProperties, "sessionProperties is null").getSessionProperties();
         this.tableProperties = requireNonNull(tableProperties, "tableProperties is null").getTableProperties();
         this.systemTables = requireNonNull(systemTables, "systemTables is null");
+        this.dao = onDemandDao(dbi, MetadataDao.class);
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: require non null dbi

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java [hunk 3]
```diff
@@ -167,7 +167,7 @@ private void discoverShards()
 
         for (Entry<Long, List<ShardMetadata>> entry : Multimaps.asMap(tableShards).entrySet()) {
             long tableId = entry.getKey();
-            if (!metadataDao.isCompactionEnabled(tableId)) {
+            if (!metadataDao.isCompactionEligible(tableId)) {
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java]: unrelated change?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 4]
```diff
@@ -89,21 +110,23 @@ public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel
     {
         checkConnectorSupports(READ_COMMITTED, isolationLevel);
         RaptorTransactionHandle transaction = new RaptorTransactionHandle();
-        transactions.put(transaction, metadataFactory.create());
+        transactions.put(transaction, metadataFactory.create(tableId -> beginDelete(tableId, transaction.getUuid())));
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: Won't this block all maintenance jobs if any queries (even reads) are running?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java [hunk 5]
```diff
@@ -276,9 +278,28 @@ long insertDistribution(
...
+            "WHERE table_id = :tableId")
+    void unblockMaintenance(@Bind("tableId") long tableId);
+
+    @SqlQuery("SELECT maintenance_blocked IS NOT NULL\n" +
+            "FROM tables\n" +
+            "WHERE table_id = :tableId\n" +
+            "FOR UPDATE")
+    boolean isMaintenanceBlockedLocked(@Bind("tableId") long tableId);
+
+    @SqlUpdate("UPDATE tables SET maintenance_blocked = NULL")
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: do we need an index for this?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 6]
```diff
@@ -52,9 +61,13 @@
     private final List<PropertyMetadata<?>> sessionProperties;
     private final List<PropertyMetadata<?>> tableProperties;
     private final Set<SystemTable> systemTables;
+    private final MetadataDao dao;
 
     private final ConcurrentMap<ConnectorTransactionHandle, RaptorMetadata> transactions = new ConcurrentHashMap<>();
 
+    @GuardedBy("this")
+    private final Multimap<Long, UUID> deletions = ArrayListMultimap.create();
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: set multimap

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 7]
```diff
@@ -167,4 +190,24 @@ public final void shutdown()
             log.error(e, "Error shutting down connector");
         }
     }
+
+    private synchronized void beginDelete(long tableId, UUID transactionId)
+    {
+        dao.blockMaintenance(tableId);
+        deletions.put(tableId, transactionId);
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: assert that this does not replace an existing transaction

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 8]
```diff
@@ -167,4 +190,24 @@ public final void shutdown()
...
+    private synchronized void finishDelete(UUID transactionId)
+    {
+        deletions.entries().stream()
+                .filter(entry -> entry.getValue().equals(transactionId))
+                .findFirst()
+                .ifPresent(entry -> {
+                    long tableId = entry.getKey();
+                    deletions.remove(tableId, transactionId);
+                    if (!deletions.containsKey(tableId)) {
+                        dao.unblockMaintenance(tableId);
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: what happens if this throws an exception?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 9]
```diff
@@ -76,6 +90,13 @@ public RaptorConnector(
         this.sessionProperties = requireNonNull(sessionProperties, "sessionProperties is null").getSessionProperties();
         this.tableProperties = requireNonNull(tableProperties, "tableProperties is null").getTableProperties();
         this.systemTables = requireNonNull(systemTables, "systemTables is null");
+        this.dao = onDemandDao(dbi, MetadataDao.class);
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        dao.unblockAllMaintenance();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: This should only happen on the coordinator

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java (hunk 8 above)
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: we should probably kick off a background job to retry this until it succeeds.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java (hunk 5 above)
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: Probably not, but I'll add one anyway. Will MySQL use it for nulls?

Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/MetadataDao.java]: It uses the index if you add a `WHERE maintenance_blocked IS NOT NULL`. Otherwise it doesn't (lame).

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java (hunk 8 above)
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: Will add a retry loop

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java (hunk 7 above)
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: Will handle this by using a `SetMultimap`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorMetadata.java (hunk 1 above)
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorMetadata.java]: Will name this `beginDeleteForTableId`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java (hunk 7 above)
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: you should still assert in that case to make sure the put had an effect

File: presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java [hunk 10]
```diff
@@ -167,4 +199,44 @@ public final void shutdown()
...
+            }
+            catch (DBIException e) {
+                log.warn(e, "Failed to unblock maintenance for table ID %s, will retry", tableId);
+            }
+
+            try {
+                SECONDS.sleep(2);
+            }
+            catch (InterruptedException e) {
+                throw Throwables.propagate(e);
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorConnector.java]: set interrupt flag
